src/hotspot/share/gc/shenandoah/heuristics/shenandoahOldHeuristics.cpp

```diff
@@ -162,21 +171,24 @@ void ShenandoahOldHeuristics::slide_pinned_regions_to_front() {
     }
   }
 
+  // Update to read from the leftmost pinned region. Plus one here because we decremented
+  // the write index to hold the next found pinned region. We are just moving it back now
+  // to point to the first pinned region.
+  _next_old_collection_candidate = write_index + 1;
+
   // Everything left should already be in the cset
-  // [ r x p p p p r ]
-  //       ^
-  //       | next pointer points at the first region which was not added
-  //       | to the collection set.
+  // [ r x x p p p p r ]
+  //   ^   ^         ^
+  //   |   |         | next pointer points at the first region which was not added to the collection set.
+  //   |   | After the above loop, the write pointer is just left of the start of the pinned regions
+  //   | Start index maintains the pointer to the beginning of the regions we scanned on this cycle.
 #ifdef ASSERT
-  for (size_t check = write_index - 1; check > _start_candidate; --check) {
+  for (size_t check = write_index; check != _start_candidate; --check) {
```

Comment:
I see that _start_candidate is initialized to zero when ShenandoahOldHeuristics is instantiated.  What I do not understand is how _start_candidate is reset to zero each time we rebuild _region_data with a new set of mixed evacuation candidates.  Can you provide a comment to explain this?

Reply:
That was a bug. I've removed this assertion because we may encounter a slot that refers to a pinned region that was moved up and that should not be in the cset.